### ISSUE TITLE
WiFi Connect not reaping child zombies

### ISSUE THREAD

**Issue by pdcastro** [balena-io/wifi-connect#361]
Not 100% sure that this is a bug or even a feature request... But maybe there is a code change that would help.

A user has a service container that runs a python script as the entry point, rather than a full init system. They reported observing `dnsmasq` zombie processes in the containers:

```
/usr/src/app# ps -Af
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  0 Oct26 pts/0    00:01:48 python3 start.py
nobody      45     1  0 Oct26 pts/0    00:00:00 [dnsmasq] <defunct>
nobody     101     1  0 Oct26 pts/0    00:00:00 [dnsmasq] <defunct>
```

One solution would be for the service container to use a init system like [tini](https://github.com/krallin/tini), but I also wonder if there would be code changes in WiFi Connect that would prevent the dnsmasq zombies. Perhaps if the WiFi Connect Node.js process "waited" on child dnsmasq processes, so they would not be inherited by the process of PID 1?

Also, I cannot rule out issues in the user's app, for example in case start.py would be issuing "kill -9" on the WiFi Connect Node.js process which would not give a chance to Node.js to wait on the children.

**Comment by pdcastro** [comment]
Connects-to: #108

**Comment by pdcastro** [comment]
@majorz, there is this `self.dnsmasq.kill();` line as you had pointed out:
https://github.com/balena-io/wifi-connect/blob/v4.4.3/src/network.rs#L196

However I see this warning in Rust's documentation:
> https://doc.rust-lang.org/std/process/struct.Child.html#warning
> On some systems, calling wait or similar is necessary for the OS to release resources. A process that terminated but has not been waited on is still around as a "zombie". Leaving too many zombies around may exhaust global resources (for example process IDs).
> The standard library does not automatically wait on child processes (not even if the Child is dropped), it is up to the application developer to do so. As a consequence, dropping Child handles without waiting on them first is not recommended in long-running applications.

**Comment by majorz** [comment]
Fixed in #362.

**Comment by majorz** [comment]
There was a merging issue - still not released. Reopening now.

**Comment by majorz** [comment]
Released as v4.4.4.
